embedded-nal-async/src/stack/tcp.rs

```diff
@@ -170,3 +170,25 @@ impl<T: TcpClientStack> TcpClientStack for &mut T {
 		T::close(self, socket)
 	}
 }
+
+/// This trait is implemented by TCP/IP stacks. In contrast to the TcpClientStack trait,
+/// this trait allows creating multiple connections using the same TCP/IP stack.
+///
+/// The trait relies on an associated type, `TcpConnection` which is returned when
+/// creating a TCP connection.
+pub trait TcpClient: embedded_io::Io {
+	/// Type returned when creating a connection.
+	type TcpConnection<'m>: embedded_io::asynch::Read + embedded_io::asynch::Write
+	where
+		Self: 'm;
+
+	/// Future returned by `connect` function.
+	type ConnectFuture<'m>: Future<Output = Result<Self::TcpConnection<'m>, Self::Error>> + 'm
+	where
+		Self: 'm;
+
+	/// Connect to the given remote host and port.
+	///
+	/// Returns `Ok` if the connection was successful.
+	fn connect<'m>(&'m mut self, remote: SocketAddr) -> Self::ConnectFuture<'m>;
```

Comment:
Hmmmm. I'm looking at the `SmolTcpClient` impl [here](https://github.com/drogue-iot/drogue-device/blob/main/device/src/drivers/tcp/smoltcp.rs#L9) and I'm no longer sure what `TcpClient` is supposed to represent after all?

The way `TcpClient` is currently implemented for smoltcp, it seems to represent a thing that - at any time - only has either zero, or one active connection? Even if we introduce lifetime `'t`, the fact that you are taking `[&mut` refs to singleton TX and RX buffers](https://github.com/drogue-iot/drogue-device/blob/main/device/src/drivers/tcp/smoltcp.rs#L48) means again - at most one active connection per `TcpClient` at a time.

So... what are we modelling with `TcpClient?
- (a) if it is a thing that can produce one active connection at a time - fine - but then we need another trait, that will model `embassy_net::Stack`. Moreover, this (`Stack`?) trait should **not** be async and  either should take a `&self` rather than `&mut self` (i.e. it should be internally synchronized), or it should be protected by a Mutex. So that I can provide references to it to e.g. the HTTP client, MQTT client and HTTP and they can all pull connections (TcpClients actually) from it "in parallel".
- (b) If it is a thing that indeed can produce multiple connections at a time, we need to think how to abstract away the notion of TX and RX buffers.

With (b) there is another problem: your current `connect` method on line 193 in this PR is async and takes `&mut` of the `TcpClient` trait. So I cannot just give out references to `TcpClient` to multiple clients and expect these to be able to pull `TcpConnections` in parallel. For that - at the very least - I'll need an async mutex. And that is if we are able to solve the TX and RX problems from above in the first place.

Reply:
Yes, a client can only have 1 connection at a time, this is on purpose. The idea is that with your stack impl, you can pre-create the number of clients your application expects so that you can do this without alloc, but you can still create the connections "on demand".

I'm not sure I see the need for having a Stack trait, because the whole point is that the application knows which stack it is using at the point of creating the clients, so it creates the number of clients == the number of connections it needs.

Example with multiple clients:

https://github.com/drogue-iot/drogue-device/blob/main/examples/nrf52/microbit/esp8266/mqtt/src/main.rs#L103-L107